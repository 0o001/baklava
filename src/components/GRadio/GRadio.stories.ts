import { storiesOf } from '@storybook/vue';
import { withKnobs, boolean } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';
import GRadio from './';
import GBox from '../GBox';
import GText from '../GText';

storiesOf('GRadio', module)
  .addDecorator(withKnobs)
  .addParameters({ component: GRadio })
  .add('Default', () => ({
    components: {
      GRadio,
      GBox,
      GText,
    },
    template: `
      <GBox>
        <GBox flex direction="row">
          <GRadio
            val="male"
            name="gender"
            v-model="selectedGender"
            :disabled="disabled"
          >
            Male
          </GRadio>
          <GRadio
            val="female"
            name="gender"
            v-model="selectedGender"
            :disabled="disabled"
          >
            Female
          </GRadio>
        </GBox>
        <GText>Selected: {{ selectedGender }}</GText>
      </GBox>
    `,
    props: {
      disabled: {
        default: boolean('Disabled', false),
      },
    },
    data () {
      return {
        selectedGender: 'female',
      };
    },
  }))
  .add('Disabled', () => ({
    components: {
      GRadio,
      GBox,
      GText,
    },
    template: `
      <GBox>
        <GBox flex direction="row">
          <GRadio
            val="male"
            name="gender"
            v-model="selectedGender"
            :disabled="disabled"
          >
            Male
          </GRadio>
          <GRadio
            val="female"
            name="gender"
            v-model="selectedGender"
            :disabled="disabled"
          >
            Female
          </GRadio>
        </GBox>
        <GText>Selected: {{ selectedGender }}</GText>
      </GBox>
    `,
    props: {
      disabled: {
        default: boolean('Disabled', true),
      },
    },
    data () {
      return {
        selectedGender: 'male',
      };
    },
  }))
  .add('Events', () => ({
    components: {
      GRadio,
      GBox,
      GText,
    },
    template: `
      <GBox>
        <GBox flex direction="row">
          <GRadio
            val="male"
            name="gender"
            @change="onMaleChange"
            v-model="selectedGender"
            :disabled="disabled"
          >
            Male
          </GRadio>
          <GRadio
            val="female"
            name="gender"
            @click="onFemaleClick"
            v-model="selectedGender"
            :disabled="disabled"
          >
            Female
          </GRadio>
        </GBox>
        <GText>Selected: {{ selectedGender }}</GText>
      </GBox>
    `,
    props: {
      disabled: {
        default: boolean('Disabled', false),
      },
    },
    methods: {
      onMaleChange (event) {
        action('onMaleChange')(event);
      },
      onFemaleClick (event) {
        action('onFemaleClick')(event);
      },
    },
    data () {
      return {
        selectedGender: '',
      };
    },
  }));
