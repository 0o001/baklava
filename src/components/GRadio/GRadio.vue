<template>
  <label class="g-radio">
    <input
      type="radio"
      v-bind="$attrs"
      v-on="listeners"
      :name="name"
      :value="val"
      :checked="isChecked"
      @click="click"
    >
    <GText
      variant="body"
      :class="{
        '-disabled': $attrs.disabled,
        '-disabled-checked': $attrs.disabled && isChecked,
      }"
    >
      <slot />
    </GText>
  </label>
</template>

<script lang="ts">

import GText from '../GText/GText.vue';

export default {
  name: 'GRadio',
  inheritAttrs: false,
  components: {
    GText,
  },
  props: {
    value: {
      type: [Boolean, String, Number, Object],
      default: '',
    },
    val: {
      type: [Boolean, String, Number, Object],
      default: '',
    },
    name: {
      type: String,
      default: '',
    },
  },
  methods: {
    click (): void {
      if (!this.$attrs.disabled && this.value !== this.val) {
        this.$emit('input', this.val);
      }
    },
  },
  computed: {
    isChecked (): boolean {
      return this.value === this.val;
    },
    listeners (): {} {
      const { input, ...rest } = this.$listeners;

      return rest;
    },
  },
};
</script>

<style lang="scss" scoped>
  .g-radio {
    cursor: pointer;
    margin-right: var(--spacing-30);

    input[type=radio] {
      display: none;

      &:checked + ::v-deep .g-text::after {
        opacity: 1;
      }

    }

    ::v-deep .g-text {
      display: inline-block;
      position: relative;
      padding-left: 26px;

      &:hover::before {
        border: 1px solid var(--orange-500);
      }

      &::before {
        content: '';
        position: absolute;
        top: var(--spacing-0);
        left: var(--spacing-0);
        width: var(--spacing-16);
        height: var(--spacing-16);
        border-radius: var(--radius-full);
        border: 1px solid var(--mid-grey-800);
        background-color: var(--white);
      }

      &::after {
        content: '';
        padding: var(--spacing-4);
        position: absolute;
        top: var(--spacing-0);
        left: var(--spacing-0);
        width: var(--spacing-10);
        height: var(--spacing-10);
        border: 4px solid var(--orange-500);
        border-radius: var(--radius-full);
        background-color: var(--white);
        opacity: 0;
      }

      &.-disabled {
        &::before {
          background-color: var(--bg-grey-500);
          border: solid 1px var(--mid-grey-800);
        }
      }

      &.-disabled-checked {
        &::after {
          content: '';
          position: absolute;
          top: var(--spacing-4);
          left: var(--spacing-4);
          width: var(--spacing-4);
          height: var(--spacing-4);
          background-color: var(--light-grey-500);
          border: 0 solid transparent;
        }
      }

    }
  }
</style>
