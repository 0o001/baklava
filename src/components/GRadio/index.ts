import GRadio from './GRadio.vue';

export default GRadio;
