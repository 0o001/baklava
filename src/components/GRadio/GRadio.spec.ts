import { mount } from '@vue/test-utils';
import GRadio from './';

describe('Radio', () => {
  let wrapper: any;
  let listenersStub = {
    input: () => {
    },
    mouseover: () => {
    },
  };
  const renderWrapper = (options?: any) => {
    wrapper = mount(GRadio, {
      propsData: {
        name: 'testName',
        value: 'Test',
        val: 'Test',
      },
      listeners: listenersStub,
      ...options,
    });
  };

  beforeEach(() => {
    renderWrapper();
  });

  it('should match with snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should match with snapshot when disabled attribute is passed', function () {
    renderWrapper({
      attrs: {
        disabled: true,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('should return true on isChecked if value and val is equal', () => {
    expect(
      wrapper.vm.isChecked
    ).toBe(true);
  });

  it('should return false on isChecked if value and val is not equal', async () => {
    wrapper.setProps({
      value: 'Test 2',
    });
    await wrapper.vm.$nextTick();

    expect(
      wrapper.vm.isChecked
    ).toBe(false);
  });

  it('should extract the input listener from listeners', () => {
    expect(
      Object.keys(wrapper.vm.listeners)
    ).toEqual(['mouseover']);
  });

  it('should emit input event if disabled is false and value and val is not equal', async () => {
    renderWrapper({
      attrs: {
        disabled: false,
      },
    });

    wrapper.setProps({
      value: 'Test 2',
    });
    await wrapper.vm.$nextTick();

    wrapper.vm.click();

    expect(wrapper.emitted().input[0][0]).toEqual(wrapper.vm.val);
  });

  it('should not emit input event if disabled is true and value and val is not equal', () => {
    renderWrapper({
      attrs: {
        disabled: true,
      },
    });

    wrapper.setProps({
      value: 'Test 2',
    });

    wrapper.vm.click();

    expect(wrapper.emitted().input).toEqual(undefined);
  });

  it('should not emit input event if disabled is false and value and val is equal', () => {
    renderWrapper({
      attrs: {
        disabled: false,
      },
    });

    wrapper.vm.click();

    expect(wrapper.emitted().input).toEqual(undefined);
  });
});
