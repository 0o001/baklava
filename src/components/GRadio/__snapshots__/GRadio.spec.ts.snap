// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio should match with snapshot 1`] = `
<label
  class="g-radio"
>
  <input
    name="testName"
    type="radio"
    value="Test"
  />
   
  <div
    class="g-text-main-grey-500 -body g-text"
  />
</label>
`;

exports[`Radio should match with snapshot when disabled attribute is passed 1`] = `
<label
  class="g-radio"
>
  <input
    disabled="disabled"
    name="testName"
    type="radio"
    value="Test"
  />
   
  <div
    class="g-text-main-grey-500 -body g-text -disabled -disabled-checked"
  />
</label>
`;
